bundles/config/org.eclipse.smarthome.config.xml/src/main/java/org/eclipse/smarthome/config/xml/AbstractXmlBasedProvider.java

```diff
@@ -0,0 +1,259 @@
+            this.locale = locale;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + getOuterType().hashCode();
+            result = prime * result + ((locale == null) ? 0 : locale.hashCode());
+            result = prime * result + ((uid == null) ? 0 : uid.hashCode());
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            @SuppressWarnings("unchecked")
+            LocalizedKey other = (LocalizedKey) obj;
+            if (!getOuterType().equals(other.getOuterType())) {
+                return false;
+            }
+            if (locale == null) {
+                if (other.locale != null) {
+                    return false;
+                }
+            } else if (!locale.equals(other.locale)) {
+                return false;
+            }
+            if (uid == null) {
+                if (other.uid != null) {
+                    return false;
+                }
+            } else if (!uid.equals(other.uid)) {
+                return false;
+            }
+            return true;
+        }
+
+        private AbstractXmlBasedProvider<T_ID, T_OBJECT> getOuterType() {
+            return AbstractXmlBasedProvider.this;
+        }
+
+    }
+
+    private final Map<Bundle, List<T_OBJECT>> bundleObjectMap = new ConcurrentHashMap<>();
+    private final Map<LocalizedKey, T_OBJECT> localizedObjectCache = new ConcurrentHashMap<>();
+
+    /**
+     * Return the unique identifier of the given object (e.g. UID, URI,...)
+     *
+     * @param object the object to return the identifier for
+     * @return the identifier. Must not be <code>null</code>
+     */
+    protected abstract T_ID getIndentifier(T_OBJECT object);
+
+    /**
+     * Create a translated/localized copy of the given object.
+     *
+     * @param bundle the bundle where the config description with its translations originates from
+     * @param object the object to translate
+     * @param locale the target locale
+     * @return a translated copy of the given object or <code>null</code> if translation was not possible.
+     */
+    protected abstract T_OBJECT localize(Bundle bundle, T_OBJECT object, Locale locale);
+
+    /**
+     * Adds an object to the internal list associated with the specified module.
+     * <p>
+     * This method returns silently, if any of the parameters is {@code null}.
+     *
+     * @param bundle the module to which the object is to be added
+     * @param object the object to be added
+     */
+    public final synchronized void add(Bundle bundle, T_OBJECT object) {
+        if (object == null) {
```

Comment:
Wouldn't it be better (with respect to removal of redundant code) to call `addAll(bundle, Collections.singletonList(object))`